File: python/2023/day13.py
from __future__ import annotations
from lib import Point, get_limits
from functools import cache, cached_property, reduce
from itertools import pairwise
from enum import Enum, auto
import operator


class Axis(Enum):
    V = auto()
    H = auto()

    def inverse(self) -> Axis:
        match self:
            case Axis.V: return Axis.H
            case Axis.H: return Axis.V


class Pt(Point):
    def pivoted(self, axis: Axis) -> Pt:
        match axis:
            case Axis.V: return self
            case Axis.H: return Pt(self.y, self.x)

    def view(self, axis: Axis) -> int:
        match axis:
            case Axis.V: return self.x
            case Axis.H: return self.y


class MirrorView(tuple):
    def __new__(cls, args) -> MirrorView:
        return tuple.__new__(cls, args)

    @cache
    def find_reflection(self, smudges: int) -> int | None:
        def is_perfect_reflection(r: int) -> bool:
            return sum(bin(a ^ b).count("1") for a, b in zip(reversed(self[:r]), self[r:])) == smudges

        for i, (ra, rb) in enumerate(pairwise(self)):
            if bin(ra ^ rb).count("1") <= smudges and is_perfect_reflection(i + 1):
                return i + 1

        return None


class Mirror(tuple):
    def __new__(cls, args) -> Mirror:
        return tuple.__new__(cls, args)

    @cache
    def parse(chunk: str) -> Mirror:
        return Mirror(
            Pt(x, y)
            for y, line in enumerate(chunk.splitlines())
            for x, ch in enumerate(line)
            if ch == "#"
        )

    @cached_property
    def extent(self) -> Pt:
        return Pt(
            get_limits(map(lambda k: k.x, self))[1],
            get_limits(map(lambda k: k.y, self))[1],
        )

    @cache
    def view(self, axis: Axis) -> MirrorView:
        return MirrorView(
            reduce(operator.or_, (
                1 << b
                for b in range(self.extent.view(axis.inverse()) + 1)
                if Pt(a, b).pivoted(axis) in self
            ))
            for a in range(self.extent.view(axis) + 1)
        )


def solve(input: str, smudges: int) -> int:
    def _solve(chunk: str) -> int:
        m = Mirror.parse(chunk)
        if v := m.view(Axis.V).find_reflection(smudges):
            return v
        elif h := m.view(Axis.H).find_reflection(smudges):
            return h * 100
        else:
            return 0

    return sum(map(_solve, input.split("\n\n")))


def part1(input: str) -> int:
    return solve(input, 0)


def part2(input: str) -> int:
    return solve(input, 1)


TEST_INPUT = """#.##..##.
..#.##.#.
##......#
##......#
..#.##.#.
..##..##.
#.#.##.#.

#...##..#
#....#..#
..##..###
#####.##.
#####.##.
..##..###
#....#..#"""

PART1_TESTS = [
    (TEST_INPUT, 405),
]

PART2_TESTS = [
    (TEST_INPUT, 400),
]

File: rust/2023/src/bin/day13.rs
#[derive(Clone, Copy)]
enum Axis {
    V,
    H,
}

impl Axis {
    fn inverse(&self) -> Axis {
        match self {
            Axis::V => Axis::H,
            Axis::H => Axis::V,
        }
    }
}

#[derive(Clone, Copy, Eq, PartialEq)]
struct Point {
    x: usize,
    y: usize,
}

impl Point {
    fn new(x: usize, y: usize) -> Point {
        Point { x, y }
    }

    fn pivoted(self, axis: Axis) -> Point {
        match axis {
            Axis::V => self,
            Axis::H => Point::new(self.y, self.x),
        }
    }

    fn view(self, axis: Axis) -> usize {
        match axis {
            Axis::V => self.x,
            Axis::H => self.y,
        }
    }
}

type MirrorView = Vec<usize>;

trait Reflection {
    fn find_reflection(&self, smudges: u32) -> Option<usize>;
}

impl Reflection for MirrorView {
    fn find_reflection(&self, smudges: u32) -> Option<usize> {
        self.windows(2).enumerate().find_map(|(i, w)| {
            if (w[0] ^ w[1]).count_ones() <= smudges
                && self[..i + 1]
                    .iter()
                    .rev()
                    .zip(self[i + 1..].iter())
                    .fold(0, |acc, (&a, &b)| acc + (a ^ b).count_ones())
                    == smudges
            {
                Some(i + 1)
            } else {
                None
            }
        })
    }
}

struct Mirror {
    points: Vec<Point>,
    extent: Point,
}

impl From<&str> for Mirror {
    fn from(chunk: &str) -> Self {
        let points = chunk
            .lines()
            .enumerate()
            .flat_map(|(y, line)| {
                line.chars().enumerate().filter_map(move |(x, c)| {
                    if c == '#' {
                        Some(Point::new(x, y))
                    } else {
                        None
                    }
                })
            })
            .collect::<Vec<_>>();

        let extent = points.iter().fold(Point::new(0, 0), |acc, r| {
            Point::new(acc.x.max(r.x), acc.y.max(r.y))
        });

        Self { points, extent }
    }
}

impl Mirror {
    fn view(&self, axis: Axis) -> MirrorView {
        (0..=self.extent.view(axis))
            .map(|a| {
                (0..=self.extent.view(axis.inverse())).fold(0, |acc, b| {
                    acc | ((self.points.contains(&Point::new(a, b).pivoted(axis)) as usize) << b)
                })
            })
            .collect()
    }
}

struct Solution {
    mirrors: Vec<Mirror>,
}

impl Solution {
    fn solve(&self, smudges: u32) -> usize {
        self.mirrors
            .par_iter()
            .fold(
                || 0,
                |acc, mirror| {
                    acc + mirror
                        .view(Axis::H)
                        .find_reflection(smudges)
                        .map(|r| r * 100)
                        .or_else(|| mirror.view(Axis::V).find_reflection(smudges))
                        .unwrap_or(0)
                },
            )
            .sum()
    }
}

impl Solver for Solution {
    fn new(input: &str) -> Anyhow<Self> {
        Ok(Self {
            mirrors: input.split("\n\n").map(Mirror::from).collect(),
        })
    }

    fn part1(&mut self) -> Anyhow<impl fmt::Display> {
        Ok(self.solve(0))
    }

    fn part2(&mut self) -> Anyhow<impl fmt::Display> {
        Ok(self.solve(1))
    }
}

aoc::solution!();

#[cfg(test)]
mod test {
    use super::{Solution, Solver};

    const INPUT: &str = "#.##..##.
..#.##.#.
##......#
##......#
..#.##.#.
..##..##.
#.#.##.#.

#...##..#
#....#..#
..##..###
#####.##.
#####.##.
..##..###
#....#..#";

    #[test]
    fn test_part1() {
        let mut solution = Solution::new(INPUT).unwrap();
        let answer = solution.part1().unwrap().to_string();
        assert_eq!(answer, "405");
    }

    #[test]
    fn test_part2() {
        let mut solution = Solution::new(INPUT).unwrap();
        let answer = solution.part2().unwrap().to_string();
        assert_eq!(answer, "400");
    }
}
